### PR TITLE
[quant][pt2] Add Conv + BN + ReLU fusion for prepare QAT

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100610
* #100525
* __->__ #100283

Summary: This follows https://github.com/pytorch/pytorch/pull/98568,
which lays all the groundwork for Conv + BN fusion in prepare QAT.
Conv + BN + ReLU fusion can reuse the same match and replace
patterns and is handled similarly.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_prepare_qat_conv_bn_relu_fusion
python test/test_quantization.py TestQuantizePT2E.test_prepare_qat_conv_bn_relu_numerics

Reviewers: kimishpatel, jerryzh168

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D45515494](https://our.internmc.facebook.com/intern/diff/D45515494)

### CODE REVIEW

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 1]
```diff
@@ -235,38 +235,71 @@ def forward(self, x):
...
+        self._verify_symmetric_qnnpack_qat_graph(M(), example_inputs, is_per_channel=True, has_relu=False)
 
-        # program capture
+    def test_prepare_qat_conv_bn_relu_fusion(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.conv = torch.nn.Conv2d(3, 3, 3)
+                self.bn = torch.nn.BatchNorm2d(3)
+                self.relu = torch.nn.ReLU()
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e.py:248]: nit: maybe you can put model definition in helper as well and you can do:
```
self.relu = torch.nn.ReLU() if has_relu else torch.nn.Identity()
```

**Review by jerryzh168 — APPROVED** [review]
LGTM, had one comment inline

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_quantize_pt2e.py:248]: actually if we want to use that function for linear - relu etc. as well it might be OK

**Comment by kimishpatel** [comment]
If you can hold off, I want to land this stack https://github.com/pytorch/pytorch/pull/99580

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 1 above)
Comment (andrewor14) [reply test/quantization/pt2e/test_quantize_pt2e.py:248]: Yeah actually I'm about to use this helper function for resnet/mobilenet numerics

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45515494).

**Pull request closed without merging**